javascript/ql/src/experimental/Security/CWE-073/TemplateObjectInjection.ql

```diff
@@ -0,0 +1,54 @@
+/**
+ * @name Template Object Injection
+ * @description Instantiating a template using a user-controlled object is vulnerable to local file read and potential remote code execution.
+ * @kind path-problem
+ * @problem.severity error
+ * @precision high
+ * @id js/template-object-injection
+ * @tags security
+ *       external/cwe/cwe-073
+ *       external/cwe/cwe-094
+ */
+
+import javascript
+import DataFlow::PathGraph
+import semmle.javascript.security.TaintedObject
+
+predicate isUsingHbsEngine() {
+  Express::appCreation().getAMethodCall("set").getArgument(1).mayHaveStringValue("hbs")
+}
```

Comment:
I tried to do a trial run of the query on 200+ projects: https://lgtm.com/query/332232496345521019/   
And it came out empty.   

(This suggestion if very optional). 

I'm thinking that this `isUsingHbsEngine()` check might not be needed for a first iteration of this query.   
There might be other template engines that have the same vulnerability, and there might be other ways of instantiating Handlebards.   
If the query turns out to be too noisy, then the check can return.

Reply:
> And it came out empty.

This is somewhat reasonable as in the run example, I see that all of them are prominent Javascript projects.

The anti-code-pattern is prone to happen in the following scenarios (Just my opinion, let me know your thoughts):

1. Project uses Server Side Rendering. 
2. This gives us the idea that the core business/production logic mostly is not present in that Express App as Server Side Rendering is mostly used for Static Sites.
3. Like frontend Node Server to just serve minimal HTML. (In my case, the bug was in this scenario.)

I think that the Github Projects or Repos which are prominent libraries or frameworks will rarely be a good target for this query.
We should go behind Repos which resemble the above-mentioned behaviors.

Reply:
> The anti-code-pattern is prone to happen in the following scenarios  

Those scenarios sound about right.   

We will see what we find when we run the query on more projects.

Reply:
Okay, I agree, let's remove that check for the test run.

Reply:
Re: removing the `isUsingHbsEngine()`, @agustingianni, which found this same bug pattern independently, found that 8/12 rendering engines were vulnerable and reported the issue to them as suggested by the Express team. So I think we are ok removing this part for now. Perhaps we can later improve the query with the list of affected engines.

Reply:
Sounds good.